### ISSUE TITLE
Load testing tool

### ISSUE THREAD

**Issue by zaach** [mozilla/picl-server#4]
We want this thing to scale, so we'll need some tools. Persona's loadgen is a good point of reference: https://github.com/mozilla/browserid/blob/dev/bin/load_gen

It runs a number of user activities in proportion to what one would expect in production. Activities are defined here: https://github.com/mozilla/browserid/tree/dev/lib/load_gen/activities

**Comment by lloyd** [comment]
Yes.  One challenge with our loadgen is I had to make up numbers.  It would probably be useful to understand from sync folks: how many users they have, how many API calls that amounts to on average, and the amount of data carried on average by each transaction.

Having this you can probably have more realistic initial load test parameters.

**Comment by zaach** [comment]
Yep, thanks lloyd. @rfk could you look into those numbers?

**Comment by rfk** [comment]
Yep.  The numbers we use for current sync loadtesting are pretty old but should be a good starting point, and I'll file a bug to get some more recent status generated

**Comment by almet** [comment]
On the services team, we've created marteau, a tool to do load-testing (http://marteau.rtfd.org). I wrote a blog post about it on the webdev blog: https://blog.mozilla.org/webdev/2013/01/18/load-testing-the-marketplace/

Maybe this can be of any use here? (cc @tarekziade)

**Comment by zaach** [comment]
Nice, we should definitely evaluate marteau.

**Comment by rfk** [comment]
Some details on the load at our SCL2 datacenter, straight from ops:
"""
Our day to day peaks for Sync hit about 1250 - 1400 QPS around 08:00 AM Pacific, and hit a trough around 4:00 PM Pacific at ~600 QPS.  We have something like twenty GET requests for every one POST requests.  The mean size for a GET request is 3471 bytes, and mean POST requests in  bytes is 213.  The median size for a GET request, on the other hand, is 252 bytes and the median POST size is just 74 bytes, and that with a median average distribution of zero.
"""

Request and response body sizes are quite variable though, with a few outliers syncing _stupefyingly huge_ numbers of history and bookmark items, that likely skew the figures.

The 20/1 ratio of GET over POST is partly due to the polling-based nature of sync, where clients must periodically hit a special "has anything changed lately?" endpoint to check whether they should sync.  I'll ask for some numbers to break this down into polling-related requests versus actual syncing-related requests.

**Comment by rfk** [comment]
During a peak load cycle last week, a 1 minute snapshot of sync saw ~18000 unique users.  We can easily get figures for the total number of users in the system, but there's been hesitation in the past to reveal this number publicly so I won't post it here.

Of the GET requests during this time, just over 50% were for the /info/collections document and therefore part of the change-polling process rather than actual syncing.  With a 20/1 ratio of GET to POST, that's a lot of polling traffic compared to useful traffic!
